crypto/serializer/deserializer_meth.c

```diff
@@ -0,0 +1,533 @@
+        OSSL_DESERIALIZER_free(method);
+    }
+}
+
+void OSSL_DESERIALIZER_do_all_provided(OPENSSL_CTX *libctx,
+                                       void (*fn)(OSSL_DESERIALIZER *deser,
+                                                  void *arg),
+                                       void *arg)
+{
+    struct deserializer_do_all_data_st data;
+
+    data.user_fn = (void (*)(void *, void *))fn;
+    data.user_arg = arg;
+    ossl_algorithm_do_all(libctx, OSSL_OP_DESERIALIZER, NULL,
+                          NULL, deserializer_do_one, NULL,
+                          &data);
+}
+
+void OSSL_DESERIALIZER_names_do_all(const OSSL_DESERIALIZER *deser,
+                                    void (*fn)(const char *name, void *data),
+                                    void *data)
+{
+    if (deser == NULL)
+        return;
+
+    if (deser->common.prov != NULL) {
+        OPENSSL_CTX *libctx = ossl_provider_library_context(deser->common.prov);
+        OSSL_NAMEMAP *namemap = ossl_namemap_stored(libctx);
+
+        ossl_namemap_doall_names(namemap, deser->common.id, fn, data);
+    }
+}
+
+const OSSL_PARAM *
+OSSL_DESERIALIZER_settable_ctx_params(OSSL_DESERIALIZER *deser)
+{
+    if (deser != NULL && deser->settable_ctx_params != NULL)
+        return deser->settable_ctx_params();
+    return NULL;
+}
+
+/*
+ * Deserializer context support
+ */
+
+/*
+ * |ser| value NULL is valid, and signifies that there is no deserializer.
+ * This is useful to provide fallback mechanisms.
+ *  Functions that want to verify if there is a deserializer can do so with
+ * OSSL_DESERIALIZER_CTX_get_deserializer()
+ */
+OSSL_DESERIALIZER_CTX *OSSL_DESERIALIZER_CTX_new(void)
+{
+    OSSL_DESERIALIZER_CTX *ctx;
+
+    if ((ctx = OPENSSL_zalloc(sizeof(*ctx))) == NULL) {
+        ERR_raise(ERR_LIB_OSSL_DESERIALIZER, ERR_R_MALLOC_FAILURE);
+        return NULL;
+    }
+
+    return ctx;
+}
+
+int OSSL_DESERIALIZER_CTX_set_params(OSSL_DESERIALIZER_CTX *ctx,
+                                     const OSSL_PARAM params[])
+{
+    size_t i;
+    size_t l;
+
+    if (!ossl_assert(ctx != NULL)) {
+        ERR_raise(ERR_LIB_OSSL_DESERIALIZER, ERR_R_PASSED_NULL_PARAMETER);
+        return 0;
+    }
+
+    if (ctx->deser_insts == NULL)
+        return 1;
+
+    l = (size_t)sk_OSSL_DESERIALIZER_INSTANCE_num(ctx->deser_insts);
+    for (i = 0; i < l; i++) {
+        OSSL_DESERIALIZER_INSTANCE *deser_inst =
+            sk_OSSL_DESERIALIZER_INSTANCE_value(ctx->deser_insts, i);
+
+        if (deser_inst->deserctx == NULL
+            || deser_inst->deser->set_ctx_params == NULL)
+            continue;
+        if (!deser_inst->deser->set_ctx_params(deser_inst->deserctx, params))
+            return 0;
+    }
+    return 1;
+}
+
+static void
+OSSL_DESERIALIZER_INSTANCE_free(OSSL_DESERIALIZER_INSTANCE *deser_inst)
+{
+    if (deser_inst != NULL) {
+        if (deser_inst->deser->freectx != NULL)
+            deser_inst->deser->freectx(deser_inst->deserctx);
```

Comment:
This assumes the freectx() handles deser_inst->deserctx == NULL?

Reply:
Yes. If I'm not mistaken, that's an assumption we've made across all provider side freeing functions.